File: backend/internal/models/note.go
package models

import (
	"time"

	"gorm.io/gorm"
)

type Note struct {
	ID        uint           `json:"id" gorm:"primaryKey"`
	Title     string         `json:"title" gorm:"not null"`
	Content   string         `json:"content"`
	IsPrivate bool           `json:"is_private" gorm:"default:true"`
	UserID    uint           `json:"user_id" gorm:"not null"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `json:"-" gorm:"index"`

	// Relationships
	User          User               `json:"user,omitempty"`
	Collaborators []NoteCollaborator `json:"collaborators,omitempty"`
}

type NoteCollaborator struct {
	ID     uint `json:"id" gorm:"primaryKey"`
	NoteID uint `json:"note_id" gorm:"not null"`
	UserID uint `json:"user_id" gorm:"not null"`

	// Relationships
	Note Note `json:"note,omitempty"`
	User User `json:"user,omitempty"`
}

type CreateNoteRequest struct {
	Title     string `json:"title" binding:"required"`
	Content   string `json:"content"`
	IsPrivate bool   `json:"is_private"`
}

type UpdateNoteRequest struct {
	Title     string `json:"title"`
	Content   string `json:"content"`
	IsPrivate *bool  `json:"is_private"`
}

type NoteResponse struct {
	ID            uint           `json:"id"`
	Title         string         `json:"title"`
	Content       string         `json:"content"`
	IsPrivate     bool           `json:"is_private"`
	UserID        uint           `json:"user_id"`
	CreatedAt     time.Time      `json:"created_at"`
	UpdatedAt     time.Time      `json:"updated_at"`
	User          UserResponse   `json:"user,omitempty"`
	Collaborators []UserResponse `json:"collaborators,omitempty"`
}

File: frontend/lib/models/note.dart
import 'package:json_annotation/json_annotation.dart';
import 'user.dart';

part 'note.g.dart';

@JsonSerializable()
class Note {
  final int id;
  final String title;
  final String content;
  final bool isPrivate;
  final int userId;
  final DateTime createdAt;
  final DateTime updatedAt;
  final User? user;
  final List<User> collaborators;

  Note({
    required this.id,
    required this.title,
    required this.content,
    required this.isPrivate,
    required this.userId,
    required this.createdAt,
    required this.updatedAt,
    this.user,
    this.collaborators = const [],
  });

  factory Note.fromJson(Map<String, dynamic> json) => _$NoteFromJson(json);
  Map<String, dynamic> toJson() => _$NoteToJson(this);

  Note copyWith({
    int? id,
    String? title,
    String? content,
    bool? isPrivate,
    int? userId,
    DateTime? createdAt,
    DateTime? updatedAt,
    User? user,
    List<User>? collaborators,
  }) {
    return Note(
      id: id ?? this.id,
      title: title ?? this.title,
      content: content ?? this.content,
      isPrivate: isPrivate ?? this.isPrivate,
      userId: userId ?? this.userId,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      user: user ?? this.user,
      collaborators: collaborators ?? this.collaborators,
    );
  }
}
